src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs

```diff
@@ -186,113 +196,132 @@ private static Delegate CreateObjectArrayDelegateRefEmit(Type delegateType, Func
 
                 if (thunkMethod == null)
                 {
-                    int thunkIndex = Interlocked.Increment(ref s_ThunksCreated);
-                    Type[] paramTypes = new Type[parameters.Length + 1];
-                    paramTypes[0] = typeof(Func<object[], object>);
-
-                    StringBuilder thunkName = new StringBuilder();
-                    thunkName.Append("Thunk");
-                    thunkName.Append(thunkIndex);
-                    if (hasReturnValue)
+                    IDisposable? forceAllowDynamicCodeScope = null;
+                    if (!RuntimeFeature.IsDynamicCodeSupported)
                     {
-                        thunkName.Append("ret_");
-                        thunkName.Append(returnType.Name);
+                        // Force 'new DynamicMethod' to not throw even though RuntimeFeature.IsDynamicCodeSupported is false.
+                        // If we are running on a runtime that supports dynamic code, even though the feature switch is off,
+                        // for example when running on CoreClr with PublishAot=true, this will allow IL to be emitted.
+                        // If we are running on a runtime that really doesn't support dynamic code, like NativeAOT,
+                        // CanEmitObjectArrayDelegate will be flipped to 'false', and this method won't be invoked.
+                        forceAllowDynamicCodeScope = ForceAllowDynamicCodeLightup.ForceAllowDynamicCodeDelegate?.Invoke();
                     }
 
-                    for (int i = 0; i < parameters.Length; i++)
+                    try
```

Comment:
Nit: use `using` to avoid try/finally? Using will not call Dispose on a null, if that was the reason for try/finally.